### ISSUE TITLE
How to call an intermediate function which calls project's main?

### ISSUE THREAD

**Issue by mcarans** [anntzer/defopt#91]
This looks like a very promising library for simplifying command line argument handling. However I have a use case which may not be handled. I read the documentation but am not sure if it is possible or not.

I was wondering how to go about calling an intermediate function (named facade) in a library that uses some arguments and then passes back to the project's main function?

eg.
```
def main(hdx_site: str = "prod", output_failures: bool = False):
    """Main function

    Args:
        hdx_site (str): HDX site
        output_failures (bool): Whetehr to output failures

    Returns:
        None
    """
    print(f"I only use output_failures which is {output_failures}")

if __name__ == "__main__":
    defopt.run(main)
```

The library has the intermediate function, facade:

```
def facade(projectmainfn: Callable[[Any], None], **kwargs: Any):
    print(f"I do some setup and use hdx_site which is {hdx_site} then I call the project's main")
    projectmainfn(**kwargs)    
```

The defopt.run line needs to use main to get the allowed command line arguments, but call facade passing main and the the parsed keyword arguments as parameters to facade. facade does some stuff using some of those arguments then calls main with them (although not all are actually used by main).

Is this possible somehow?

**Comment by mcarans** [comment]
@anntzer I assume you meant `="kwonly"` (strict_kwonly=True) above.

From what I can tell from looking through the code, it is this logic that needs to be altered for the `options` flag:
```
    positionals = {name for name, param in sig.parameters.items()
                   if ((param.default is param.empty or strict_kwonly)
                       and not _is_list_like(param.annotation)
                       and param.kind != param.KEYWORD_ONLY)}
```
I think it should be something like the below if I've understood your suggested `options` flags correctly:
```
    positionals = {name for name, param in sig.parameters.items()
                   if (((param.default is param.empty and options == 'has_default') or options == 'kwonly')
                       and not _is_list_like(param.annotation)
                       and param.kind != param.KEYWORD_ONLY)}
```

If this looks like the right kind of approach to you, I'll have a look at it further over the coming days.

**Comment by anntzer** [comment]
Yes.  Note that in order to correctly deprecate strict_kwonly you'll need to do something like
```python
_unset = object()
def run(...,
        options=_unset,
        strict_kwonly=_unset,
        ...):
    ...
    if options is not _unset and strict_kwonly is not _unset:
        raise ValueError("Cannot pass both 'options' and 'strict_kwonly'")
    if strict_kwonly is not _unset:
        warnings.warn(DeprecationWarning, ...)
        options = "kwonly" if strict_kwonly else "has_default"

run.__signature__ = <signature with only options="has_default", without strict_kwonly>
```

**Comment by mcarans** [comment]
@anntzer I've made a PR: https://github.com/anntzer/defopt/pull/92

**Comment by anntzer** [comment]
Closed by #92.
